execution/stream_join.go

```diff
@@ -0,0 +1,448 @@
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't read single records from record set")
+		}
+
+		for i := range singleRecords {
+			singleRecord := valueToRecord(singleRecords[i])
+			idForRecord := NewRecordIDFromStreamIDWithOffset(js.streamID, triggeredCountValue+i)
+			WithID(idForRecord)(singleRecord)
+
+			records = append(records, singleRecord)
+		}
+
+		if err := recordsToTriggerSet.Clear(); err != nil {
+			return nil, errors.Wrap(err, "couldn't clear the set of single records")
+		}
+
+		newTriggeredCount = octosql.MakeInt(triggeredCountValue + len(singleRecords))
+
+		if err := triggeredCountState.Set(&newTriggeredCount); err != nil {
+			return nil, errors.Wrap(err, "couldn't update count of triggered records after triggering single records")
+		}
+	}
+
+	return records, nil
+}
+
+func keepTrackOfCount(tx storage.StateTransaction, isRetraction bool) (int, error) {
+	var recordCount octosql.Value
+	recordCountState := storage.NewValueState(tx)
+
+	err := recordCountState.Get(&recordCount)
+	if err == storage.ErrNotFound {
+		recordCount = octosql.MakeInt(0)
+	} else if err != nil {
+		return 0, errors.Wrap(err, "couldn't check the count of record")
+	}
+
+	var updatedRecordCount octosql.Value
+	if isRetraction {
+		updatedRecordCount = octosql.MakeInt(recordCount.AsInt() - 1)
+	} else {
+		updatedRecordCount = octosql.MakeInt(recordCount.AsInt() + 1)
+	}
+
+	// Store the new count
+	if updatedRecordCount.AsInt() == 0 {
+		err = recordCountState.Clear()
+		if err != nil {
+			return 0, errors.Wrap(err, "couldn't clear count state of record")
+		}
+	} else {
+		err = recordCountState.Set(&updatedRecordCount)
+		if err != nil {
+			return 0, errors.Wrap(err, "couldn't update the count of record")
+		}
+	}
+
+	return updatedRecordCount.AsInt(), nil
+}
+
+func recordToValue(rec *Record) octosql.Value {
+	fields := make([]octosql.Value, len(rec.FieldNames))
+	data := make([]octosql.Value, len(rec.Data))
+	eventTimeField := octosql.MakeString(rec.EventTimeField().String())
+
+	for i := range rec.FieldNames {
+		fields[i] = octosql.MakeString(rec.FieldNames[i])
+		data[i] = *rec.Data[i]
+	}
+
+	valuesTogether := make([]octosql.Value, 4)
+
+	valuesTogether[0] = octosql.MakeTuple(fields)
+	valuesTogether[1] = octosql.MakeTuple(data)
+	valuesTogether[2] = octosql.MakeBool(rec.IsUndo())
+	valuesTogether[3] = eventTimeField
+
+	return octosql.MakeTuple(valuesTogether)
+}
+
+func valueToRecord(val octosql.Value) *Record {
+	slice := val.AsSlice()
+
+	fieldNameValues := slice[0].AsSlice()
+	values := slice[1].AsSlice()
+	isUndo := slice[2].AsBool()
+	eventTimeField := octosql.NewVariableName(slice[3].AsString())
+
+	fieldNames := make([]octosql.VariableName, len(fieldNameValues))
+
+	for i, fieldName := range fieldNameValues {
+		fieldNames[i] = octosql.NewVariableName(fieldName.AsString())
+	}
+
+	if isUndo {
+		return NewRecordFromSlice(fieldNames, values, WithEventTimeField(eventTimeField), WithUndo())
+	}
+
```

Comment:
I think you can remove this empty line, because return values are directly impacted by this if statement (and only this statement) - anyways, do as you think, just a suggestion